docs/proposals/image-prepull.md

```diff
@@ -0,0 +1,293 @@
+type ImagePrePullSpec struct {
+  // ImagePrePullTemplate represents original templates of imagePrePull
+  ImagePrePullTemplate ImagePrePullTemplate `json:"imagePrePullTemplate,omitempty"`
+  
+  // ImagePrePullScope represents the override rules that would apply on imagePrePull job.
+  // Will support in next version, more details in [Plan](#Plan) 
+  //ImagePrePullScope []ImagePrePullOverrider `json:"imagePrepullScope,omitempty"`
+}
+
+  // ImagePrePullTemplate represents original templates of imagePrepull
+type ImagePrePullTemplate struct {
+  // Images is the image list to be prepull
+  Images []string `json:"images`
+  
+  // NodeNames is a request to select some specific nodes. If it is non-empty,
+  // the upgrade job simply select these edge nodes to do upgrade operation.
+  // Please note that sets of NodeNames and LabelSelector are ORed.
+  // Users must set one and can only set one.
+  // +optional
+  NodeNames []string `json:"nodeNames,omitempty"`
+  // LabelSelector is a filter to select member clusters by labels.
+  // It must match a node's labels for the NodeUpgradeJob to be operated on that node.
+  // Please note that sets of NodeNames and LabelSelector are ORed.
+  // Users must set one and can only set one.
+  // +optional
+  LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+  // ParallelImagePulls sets the number of image pulls in parallel.
+  // It defaults to 1
+  // +optional
+  ParallelImagePulls *int32 `json:"parallelImagePulls,omitempty"`
+
+  // CheckItems specifies the items need to be checked before the task is executed.
+  // The default CheckItems value is disk.
+  // +optional
+  CheckItems []string `json:"checkItems,omitempty"`
+  
+  // ImageSecret specifies the secret for image pull if private registry used. 
+  // Use {namespace}/{secretName} in format. 
+  // +optional
+  ImageSecret string `json:"imageSecrets,omitempty"`
+
+  // TimeoutSecondsOnEachNode limits the duration of the image prepull job on each edgenode.
+  // Default to 360.
+  // If set to 0, we'll use the default value 360.
+  // +optional
+  TimeoutSecondsOnEachNode *uint32 `json:"timeoutSecondsOnEachNode,omitempty"`
+  
+  // RetryTimes specifies the retry times if image pull failed on each edgenode.
+  // Default to 0
+  // +optional
+  RetryTimes *uint32 `json:"retryTimes,omitempty"`
+}
+
+// ImagePrePullJobStatus stores the status of ImagePrePullJob.
+// contains images prepull status on multiple edge nodes.
+type ImagePrePullJobStatus struct {
+   // State represents for the state phase of the ImagePrePullJob.
+   // There are four possible state values: "", prepulling, successful and failed.
+   State string `json:"state,omitempty"`
+  
+  // Status contains image prepull status for each edge node.
+   Status []ImagePrePullStatus `json:"status,omitempty"`
+}
+
+// ImagePrePullStatus stores image prepull status for each edge node.
+type ImagePrepullStatus struct {
```

Comment:
The latest ImagePrePullStatus structure is:
```golang
type ImagePrePullStatus struct {
	// TaskStatus represents the status for each node
	*TaskStatus `json:"nodeStatus,omitempty"`
	// ImageStatus represents the prepull status for each image
	ImageStatus []ImageStatus `json:"imageStatus,omitempty"`
}
```
The memeber TaskStatus refers to the structure of the node upgrade task. How will its fields(status, event, action, reason, time) be represented in the image pre pull task?

Reply:
TaskStatus is not specifically refer to node upgrade task, it's an abstracted status for all tasks. In Image Prepull task, status means the overall state of the pull task, event represents specific events in images prepull process, such as check, pull, and etc.

Reply:
> TaskStatus is not specifically refer to node upgrade task, it's an abstracted status for all tasks. In Image Prepull task, status means the overall state of the pull task, event represents specific events in images prepull process, such as check, pull, and etc.

Okey, I think it will improve readability if we define upgrade node task constants and image prepull task constants completely separately, although they will have partially the same value.